core/modules/replica/Configuration.cc

```diff
@@ -23,818 +23,1732 @@
+        if (databaseJson.at("family_name").get<string>() == name) {
+            databasesToBeRemoved.push_back(database);
+        }
+    }
+    for (string const& database: databasesToBeRemoved) {
+        _data["databases"].erase(database);
+    }
+    _data["database_families"].erase(name);
 }
 
 
 size_t Configuration::replicationLevel(string const& family) const {
-    util::Lock lock(_mtx, context(__func__));
-    return _impl->replicationLevel(family);
-}
-
-
-vector<string> Configuration::databases(string const& family, bool allDatabases, bool isPublished) const {
-    util::Lock lock(_mtx, context(__func__));
-    return _impl->databases(family, allDatabases, isPublished);
+    util::Lock const lock(_mtx, _context(__func__));
+    return _databaseFamily(lock, family).at("min_replication_level").get<size_t>();
+}
+
+
+vector<string> Configuration::databases(
+        string const& family, bool allDatabases, bool isPublished) const
+{
+    util::Lock const lock(_mtx, _context(__func__));
+    if (!family.empty()) _assertDatabaseFamilyExists(lock, family, true);
+    vector<string> names;
+    for (auto&& itr: _data.at("databases").items()) {
+        string const& name = itr.key();
+        json const& obj = itr.value();
+        if (!family.empty() && (family != obj.at("family_name").get<string>())) {
+            continue;
+        }
+        if (!allDatabases) {
+            bool const databaseIsPublished = obj.at("is_published").get<int>() != 0;
+            if (isPublished != databaseIsPublished) continue;  
+        }
+        names.push_back(name);
+    }
+    return names;
 }
 
 
 bool Configuration::isKnownDatabase(string const& name) const {
-    util::Lock lock(_mtx, context(__func__));
-    return _impl->isKnownDatabase(name);
+    util::Lock const lock(_mtx, _context(__func__));
+    if (name.empty()) throw invalid_argument(_context(__func__) + " the database name is empty.");
+    return _data.at("databases").count(name) != 0;
 }
 
 
 DatabaseInfo Configuration::databaseInfo(string const& name) const {
-    util::Lock lock(_mtx, context(__func__));
-    return _impl->databaseInfo(name);
+    util::Lock const lock(_mtx, _context(__func__));
+    return DatabaseInfo(_database(lock, name));
 }
 
 
 DatabaseInfo Configuration::addDatabase(DatabaseInfo const& info) {
-    util::Lock lock(_mtx, context(__func__));
-    return _impl->addDatabase(info);
+    util::Lock const lock(_mtx, _context(__func__));
+    _assertDatabaseFamilyExists(lock, info.family, true);
+    _assertDatabaseExists(lock, info.name, false);
+    if (_connectionPtr != nullptr) {
+        _connectionPtr->executeInOwnTransaction([&](decltype(_connectionPtr) conn) {
+            auto const isNotPublished = 0;  // the new database can't be published at this time
+            conn->executeInsertQuery("config_database", info.name, info.family, isNotPublished,
+                                     info.chunkIdColName, info.subChunkIdColName);
+        });
```

Comment:
This looks like a general pattern for many methods - you first verify that operation is consistent with your in-memory state and then try modify database contents but you don't verify that database contents is still consistent with in-memory state. This sort of implies that there is no concurrency at all and database state is always the same as your in memory state?

Reply:
In the present model, all modifications of the configuration are made via the Master Replication Controller (there is the built-in REST server that supports this type of reconfigurations). In theory, one may go directly into the MySQL database and modify the configuration parameters directly. I doubt this has a good solution. Similar things may happen in Qserv as well.
As a matter of fact, the transient configuration is a cache of the persistent one. The cache is needed due to high traffic to the Configuration from many threads and clients within the Replication/Ingest system. The consistency of the cache within the Master Controller is guaranteed by a single instance of the Configuration service within the Controller and by the synchronized implementation of the service's API. And for workers, the cache consistency is enforced by the Controller-workers protocol in which the Controller may request workers to update their state on major events that resulted in configuration changes (changes made to the general parameters, collection of workers, database families, or databases).
Note that this philosophy was in the original design of the Replication/Ingest system, and it was not revised with this refactoring effort. The "new" code you're seeing has moved from one of the older classes. If you think we should redesign the service then we could discuss it at a group meeting.

Reply:
Don't you also have ConfigApp that updates database directly, not using Master Controller (this is what I thought it does but I may be wrong)?

Reply:
If the Controller is not being run then one can use {{ConfigApp}} to inspect and update the database. Both use the same API provided by the class {{Configuration}}. The application could be integrated into the container start-up sequences in the Kubernetes-based deployments. Right now Fabrice directly modifies the corresponding table(s) in MySQL. He would need to switch to using this application to avoid being directly dependent on the MySQL schema. The configuration tool is meant to help with that.